### PR TITLE
Optdriver

### PR DESCRIPTION
**Pull request by PeytonRoden**
## Description
Add functionality to do combinatorial optimization by converting the problem to a qubo problem and using a factorization machine and simulated annealing to predict minimas. 
Add functionality to do optimize thickness of layers and alloy percentage of layers by converting the problem to a qubo problem and using a factorization machine and simulated annealing to predict minimas. 

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - added function optimize_qubo(fom_func), the user can add two options to args for doing different types of optimization. The user can specify either  "combinatorial_optimization": True or  "qubo_thickness_optimization": True  in args, depending on what type of optimization they are looking for. If the user wants to do either of these types of optimizations they must pass in a dictionary to args describing how the optimization will be performed.  Users also pass in a function, fom_func, to calculate the figure of merit they want to minimize. 
 
  - Added class _qubo_thickness_and_alloy_optimization:
 This class will parse the optimization_dictionary, generate random bitstrings as training data that correspond to thickness ranges and  different alloy combinations, create a factorization machine model, train it on the training data, convert the model to a bqm and predict the minimum using simulated annealing. The minimum is added back to the training data and the other steps repeat continuing to predict better minima as more data is generated. 
  - Added function _qubo_thickness_and_alloy_optimization.__init__():
  This function parses out the dictionary that describes the optimization into some other dictionaries that are used by the other methods to covert the bitstring back to a structure and convert its figure of merit.
  - Added function _qubo_thickness_and_alloy_optimization.random_binary( self, length, num):
  This function generates random binary bitstrings to be used as the training data
   - Added function _qubo_thickness_and_alloy_optimization.binary_to_int(self, bin):
This function converts binary strings to integers
  - Added function _qubo_thickness_and_alloy_optimization.MaxwellGarnett(self, ri_1, ri_2, fraction):
  a function that will compute the alloy refractive index between material_1 and material_2 using Maxwell-Garnett theory.
  - Added function _qubo_thickness_and_alloy_optimization.learning_loop()
  This function generates the random bitstrings and converts them to structures using the update_multilayer function next the loop starts that trains the factorization machine, predicts minima, and adds them back to training data
  - Added function _qubo_thickness_and_alloy_optimization.update_multilayer()
  This function takes a bitstring parses it out to a structure and calculates that structures figure of merit
  - Added function _qubo_thickness_and_alloy_optimization.update_multilayer()
  This function takes a bitstring parses it out to a structure and calculates that structures figure of merit
  - Added function _qubo_thickness_and_alloy_optimization.convert_qubo_result_to_df()
this function takes a minima predicted by the simulated annealing and converts to a bitstring and structure to be added back to the training data.
  - Added function _qubo_thickness_and_alloy_optimization.convert_qubo_results_to_df()
this function takes multiple minima predicted by the simulated annealing and converts to bitstrings and structure to be added back to the training data.

  - Added class _qubo_combinatorial_structure_optimization:
  This class will parse the optimization_dictionary, generate random bitstrings as training data that correspond to different combinations of materials defined in the optimization dictionary, create a factorization machine model, train it on the training data, convert the model to a bqm and predict the minimum using simulated annealing. The minimum is added back to the training data and the other steps repeat continuing to predict better minima as more data is generated. 
  - Added function _qubo_combinatorial_structure_optimization.__init__():
  This function will parse the optimization dictionary into a list of methods that swap materials from the multilayer structure and a list of the materials. 
  - Added function _qubo_combinatorial_structure_optimization.random_binary():
  This function creates some random binary strings as training data
  - Added function  _qubo_combinatorial_structure_optimization.learning_loop()
  This function creates random training data as bitstrings and evaluates those bitstrings as structures and calculates an fom for each one. then the loop starts that trains the factorization machine, predicts minima, and adds them back to training data
  - Added function _qubo_combinatorial_structure_optimization.update_multilayer()
  This function takes a bitstring parses it out to a structure and calculates that structures figure of merit
  - Added function _qubo_combinatorial_structure_optimization.convert_qubo_result_to_df()
this function takes a minima predicted by the simulated annealing and converts to a bitstring and structure to be added back to the training data.

Added class _FMTrain
inside this class is another class _TorchFM which is a pytorch factorization machine model that can be trained
_FMTrain allows the factorization machine model to be trained and also allows the FM model to be converted to a bqm model and use simulated annealing to predict a minimum
  - Added function _FMTrain.__init__():
    initializes the factorization machine with variables n and k
  - Added function _FMTrain.train():
 trains the factorization machine using the provided training data and parameters
  - Added function _FMTrain.eval():
can be used to evaluate the inference accuracy of the factorization machine
  - Added function _FMTrain.get_weights():
 get the weights of the factorization machine for building the hamiltonian
  - Added function _FMTrain.return_weights():
 get the weights of the factorization machine for building the hamiltonian
  - Added function _FMTrain.triu_mask():
 creates an upper triangular matrix
- added function _FMTrain.VtoQ()
converts V matrix to Q matrix for bqm model basically just does does product of V and V.t and multiplies this by an upper triangular matrix
- added function _FMTrain.get_qubo_hamiltonian_minima():
builds the hamiltonian from the factorization machine and predicts minimum
- added function _FMTrain.get_qubo_hamiltonian_minimas():
builds the hamiltonian from the factorization machine and predicts multiple minima
- added function _FMTrain.predict(x):
get prediction from factorization machine

## Questions
- [ ] How to implement unit testing
## Status
- [x] Ready to go

### CODE REVIEW

**Comment by FoleyLab** [comment]
@PeytonRoden Although OptDriver doesn't have unit tests, it did break the old unit tests I think because it requires additional dependencies.  I can add the dependencies to devtools/conda-envs/test_env.yaml but before I do, can you review the import statements and let me know if all of them are required for the functionality of the library in its current state?

I expect we will need torch and pyqubo.  However, I am curious if any of the following are really needed or if they were used during testing and experimentation in the development process?

matplotlib, plotly, tqdm, sklearn, copy, neal, math, pandas

**Comment by PeytonRoden** [comment]
I removed matplotlib, plotly, and math dependencies. I don't know if it would be easy to remove the other dependencies.

File: wptherml/optdriver.py [hunk 1]
```diff
@@ -0,0 +1,1205 @@
...
+        if "upper_bound" in args:
+            self.upper_bound = args["upper_bound"]
+        else:
+            self.upper_bound = 1000
+
+
+        if "combinatorial_optimization" in args:
+            self.combinatorial_optimization = True
+            self.qubo_thickness_optimization = False
+            self.optimization_dictionary = args['optimization_dictionary']
```
Comment (FoleyLab) [inline wptherml/optdriver.py]: @PeytonRoden I have a few thoughts about optimization_dictionary: (1) can the entries of this dictionary just fit into the dictionary that the user passes as args?  (2) we need to figure an easy and comprehensive way to provide documentation for the user arguments because they are numerous now.  In principle Sphinx can provide documentation from docstrings that could be included in the parse_input(), parse_optimization_input(), and parse_therml_input() methods... that might be the way to go, we should look into this.

File: wptherml/optdriver.py [hunk 2]
```diff
@@ -0,0 +1,1204 @@
...
+                arr.append(list(np.random.choice([0, 1], size=(length,))))
+            return arr
+
+        def binary_to_int(self, bin):
+            #convert binary to an integer
+            num = int("".join(str(x) for x in bin), 2)
+            return num
+
+
+        def MaxwellGarnett(self, ri_1, ri_2, fraction):
```
Comment (FoleyLab) [inline wptherml/optdriver.py]: I am think we should move MaxwellGarnett() to materials.py

File: wptherml/optdriver.py [hunk 3]
```diff
@@ -0,0 +1,1204 @@
...
+            grad *= -1e-9
+
+        print(F' Current FOM is {fom}')
+        print(F' Current Gradient is {grad}')
+        return fom, grad
+
+
+    class _qubo_thickness_and_alloy_optimization:
+
+        def __init__(self, outer_instance, optimization_dict: dict, fom_func):
```
Comment (FoleyLab) [inline wptherml/optdriver.py]: If I understand correctly, the way the constructor works, the user will need to define a fom_func outside of calling optdriver.  Ideally, the syntax of computing the fom should be hidden from the user and we should have a few methods in optdriver and the qubo classes like the compute_fom_and_gradient_from_thickness_array method that will be called by the appropriate optimization driver (bfgs, qubo, etc).  Right now I have only one fom in that compute_fom_... method, but I will add more conditionals based on the value of self.objective_function that will compute different foms and gradients (when available).  Can you make this adaptation to the _qubo classes, as well?

File: wptherml/optdriver.py [hunk 4]
```diff
@@ -0,0 +1,1204 @@
...
+
+            W = np.array(np.array(params[1])[0], dtype = np.float64)
+
+            bias = np.array(params[0], dtype = np.float64)[0]
+
+
+            Q = V @ V.T
+
+
+            Q = self.VtoQ(V)
```
Comment (FoleyLab) [inline wptherml/optdriver.py]: Q = V @ V.T on line 1073 is what I would do based on the math from the different papers we have read about mapping FM to QUBO, do you know why multiplying by the upper triangle matrix (triu) is necessary in the VtoQ method?  

Also right now you are computing Q twice, once in line 1073 and again in 1076

File: wptherml/optdriver.py [hunk 5]
```diff
@@ -0,0 +1,1204 @@
...
+            
+            #if self.verbose: print(sampleset)
+            decoded_samples = model.decode_sampleset(sampleset)
+            best_sample = min(decoded_samples, key=lambda x: x.energy)
+
+
+            return best_sample
+
+        
+        def get_qubo_hamiltonian_minimas(self, num_samples):
```
Comment (FoleyLab) [inline wptherml/optdriver.py]: Could get_qubo_hamiltonian_minimas and get_qubo_hamiltonian_minima be simplified into one method that determines if one or multiple minima should be computed based on user input?

**Review by FoleyLab — CHANGES_REQUESTED** [review]
I have added some comments to optdriver.py on some changes to make and also some things to think about and potentially make changes.  Let me know if you would like to discuss any of these comments further!
